metadata_proxy.go

```diff
@@ -56,11 +69,30 @@ func (x xForwardedForStripper) RoundTrip(req *http.Request) (*http.Response, err
 	return http.DefaultTransport.RoundTrip(req)
 }
 
+// responseWriter wraps the given http.ResponseWriter to record metrics.
+type responseWriter struct {
+	filterResult string
+	http.ResponseWriter
+}
+
+func newResponseWriter(rw http.ResponseWriter) *responseWriter {
+	return &responseWriter{
+		"",
+		rw,
+	}
+}
+
+// WriteHeader records the header and writes the appropriate metric.
+func (m responseWriter) WriteHeader(code int) {
+	metrics.RequestCounter.WithLabelValues(m.filterResult, strconv.Itoa(code)).Inc()
+	m.ResponseWriter.WriteHeader(code)
+}
+
 type metadataHandler struct {
 	proxy *httputil.ReverseProxy
 }
 
-func newMetadataHandler() http.Handler {
+func newMetadataHandler() *metadataHandler {
```

Comment:
Why do you need to specify type here?

Reply:
No immediate reason, but good practice to accept interfaces, return concrete types.  https://www.reddit.com/r/golang/comments/5cos3x/accept_interfaces_return_structs/